prov/rxm/src/rxm_cq.c

```diff
@@ -512,20 +514,23 @@ static ssize_t rxm_cq_read(struct fid_cq *msg_cq, struct fi_cq_msg_entry *comp)
 	return rxm_cq_report_error(util_cq, &err_entry);
 }
 
-void rxm_cq_progress(struct fid_cq *msg_cq)
+void rxm_cq_progress(struct rxm_ep *rxm_ep)
 {
-	struct fi_cq_msg_entry comp;
-	ssize_t ret = 0;
+	struct fi_cq_tagged_entry comp;
+	ssize_t ret, comp_read = 0;
 
 	do {
-		ret = rxm_cq_read(msg_cq, &comp);
+		ret = rxm_cq_read(rxm_ep->msg_cq, &comp);
 		if (ret < 0)
 			goto err;
 
-		ret = rxm_cq_handle_comp(&comp);
-		if (ret)
-			goto err;
-	} while (ret > 0);
+		if (ret) {
+			comp_read += ret;
+			ret = rxm_cq_handle_comp(rxm_ep, &comp);
+			if (ret)
+				goto err;
+		}
+	} while (comp_read > 0 && comp_read < rxm_ep->comp_per_progress);
```

Comment:
If you read one completion (comp_read = 1), then no more are available, does this loop spin?  I'm wondering if the comp_read > 0 check is what's wanted, or if it should be ret > 0